pkgs/top-level/all-packages.nix

```diff
@@ -16103,7 +16103,10 @@ with pkgs;
   };
   swiPrologWithGui = swiProlog.override { withGui = true; };
 
-  tbb = callPackage ../development/libraries/tbb { };
+  tbb_2020_3 = callPackage ../development/libraries/tbb/2020_3.nix { };
+  tbb_2021_8 = callPackage ../development/libraries/tbb { };
+  # many packages fail with latest version, so they need to be updated one-by-one
+  tbb = tbb_2020_3;
```

Comment:
It's a reasonable change. I have only one complain on the `one-by-one` bit: updating users `one-by-one` is not really an option for library users due to diamond dependency problems: https://trofi.github.io/posts/248-how-do-shared-library-collisions-break.html. We will likely need to update most of users in a single go later. Otherwise we will have to debug non-trivial linkage failures or even plain SIGSEGVs.

Reply:
About 100 packages build and 100 fail (https://github.com/NixOS/nixpkgs/pull/214762#issuecomment-1418286637). So i thought we could change the working 100 to the new version and leave the other on the old.

You say that will cause unexpected issues? I did not have time to read the post.

So far i understand that when we have 2 libraries, that use the two versions of tbb and one package that use these two libraries, we will have issues. So i would try to update as much as possible and test the program. I guess in case of an issue, they will not even start?

How does this contradict the feature of Nix to be able to have multiple versions of libraries on the same system? Why is it different here than with other dependencies? I guess it would also not be possible to have multiple glibc? I hope such limitations of Nix are documented somewhere.

The whole point of this change (for me) is to be able to use the new version for the [CCTag](https://github.com/alicevision/CCTag) library, that is needed for [Meshroom](https://github.com/NixOS/nixpkgs/pull/216403).

Reply:
What change do you suggest here?

Reply:
> So far i understand that when we have 2 libraries, that use the two versions of tbb and one package that use these two libraries, we will have issues. So i would try to update as much as possible and test the program. I guess in case of an issue, they will not even start?

Yeah, if a single address space loads multiple library versions there will likely be troubles. It's hard to say how those will manifest. At least you will get an early crash. At worst it will not crash at all and will corrupt your data. The typical problem is when data structure is constructed in one library version and processed in another. I usually use `lddtree -a` on a binary to see if there are multiple versions of a library loaded together.

> How does this contradict the feature of Nix to be able to have multiple versions of libraries on the same system? Why is it different here than with other dependencies? I guess it would also not be possible to have multiple glibc? I hope such limitations of Nix are documented somewhere.

You can have multiple versions of `glibc` as long as you guarantee that a single address space (a single process) does not load two versions of `glibc` at the same time. `nixpkgs` tries very hard to do it by defining a single `glibc` package used everywhere (except the bootstrap). That way as long as you don't mix `glibc` from different `nixpkgs` instances in a single derivation it should be mostly OK (mostly because there are some exceptions like `/run/opengl-driver` that expose a particular `glibc` version via shared plugin, https://github.com/nixpkgs-architecture/issues/issues/15).

Mixing and matching final binaries from different `nixpkgs` version normally is safer as each of them usually uses their consistent (either disjoint or identical) closure of shared libraries and configs. There are still creative ways of breaking them like incompatible configs in `/etc/` or through other non-hermetic paths like plugins or incompatible database formats in `/var/.

The libraries are just a lot simpler to break if a single `nixpkgs` provides multiple versions of a library and uses different versions of a library in different subtrees of a package tree. `qt` (for example) tries to solve the mixing problem by using a package set that pin the whole set of `qt` libraries into a consistent set (`libsForQt5.callPackage` instead of `callPackage`). It's still a mess if your package want to use both `qt5` and `qt6` at the same time (directly or indirectly).

I suggest avoiding `one-by-one` phrasing: ` # many packages still fail with latest version`.

Reply:
> if a single address space loads multiple library versions there will likely be troubles

can we solve that by showing an error when evaluating a package that has dependencies that use the same library in different versions? it would be hard to know which package is a library and which is the same library, but in a different version. nix does not even have a concept of a package

we could have metadata for libraries like using `outputs = [ "lib" "dev" ];` and define lib as default (out) and a policy to use the same package name and only use different version, so nix can compare them

since nix already knows all dependencies, i think it could be solvable at evaluation time

Reply:
I think heuristics are possible but might be non-trivial to implement:

1. `nix` / `nixpkgs` does not exactly know it's runtime closures unti the package is built. At eval time we only see built depends in `buildInputs` / `nativeBuildInputs` and even that is both a too wide list and an incomplete list. Things like `{ foo }: stdenv.mkDerivation { .., postInstall = "echo LD_PRELOAD=${foo}/lib/libfoo.so.1 >> something"; }` don't have the `foo` in `*Inputs` as an expression, but does have it as a derivation input.

2. There are rare cases when we mix libraries, like cross-compilation, bootstrap and copied libraries (`libgcc_s.so` is provided by `gcc` and `glibc`).